### PR TITLE
Add  com.carpeludum.KegaFusion

### PR DESCRIPTION
**Pull request by hadess**
Unmaintained, i386-only emulator.

Plays Mega Drive and Master System like (the already packaged) BlastEm. It also plays Mega-CD, 32X, Game Gear, SG-1000, SC-3000.

TODO:
- Update to 19.08 runtime (there's no i386 support, so how do we compile/run for i386?)
- ~~Sound doesn't work:~~ (Fixed now)
```
ALSA lib ../../src/dlmisc.c:170:(snd_dlsym_verify) unable to verify version for symbol _snd_pcm_plug_open
ALSA lib ../../src/dlmisc.c:296:(snd1_dlobj_cache_get) symbol _snd_pcm_plug_open is not defined inside [builtin]
```

### CODE REVIEW

File: com.carpeludum.KegaFusion.metainfo.xml [hunk 1]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- Copyright 2018 Bastien Nocera <hadess@hadess.net> -->
+<component type="desktop">
```
Comment (bilelmoussaoui) [inline com.carpeludum.KegaFusion.metainfo.xml]:
```suggestion
<component type="desktop-application">
```

**Comment by bilelmoussaoui** [comment]
bot, build com.carpeludum.KegaFusion

File: com.carpeludum.KegaFusion.json [hunk 2]
```diff
@@ -0,0 +1,86 @@
+{
+    "app-id" : "com.carpeludum.KegaFusion",
+    "runtime" : "org.freedesktop.Platform",
+    "runtime-version" : "18.08",
```
Comment (LeandroStanger) [inline com.carpeludum.KegaFusion.json]: Update runtime to 19.08

Comment (hadess) [reply com.carpeludum.KegaFusion.json]: The first comment literally explains why I can't do that.

Incidentally, it's starting to get seriously annoying seeing your only interactions with me and the software that I maintain be "upgrade to this runtime version" with no explanations as to what it would entail, not thinking about the work it would take, and not answering when I ask those.

File: com.carpeludum.KegaFusion.desktop [hunk 3]
```diff
@@ -0,0 +1,10 @@
+[Desktop Entry]
+Name=Kega Fusion
+Comment=Sega 8 and 16-bit consoles emulator
+Type=Application
+MimeType=application/x-sega-cd-rom;application/x-sega-pico-rom;application/x-genesis-rom;application/x-genesis-32x-rom;application/x-sg1000-rom;application/x-sms-rom;application/x-gamegear-rom;
+Categories=Audio;AudioVideo;Graphics;GTK;Player;Video;Viewer;
```
Comment (gasinvein) [inline com.carpeludum.KegaFusion.desktop]: All this categories seems irrelevant. Is it a left-over?
```suggestion
Categories=Game;Emulator;
```

File: com.carpeludum.KegaFusion.json [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                "install -d /app/share/doc/KegaFusion",
+                "cp *.txt /app/share/doc/KegaFusion",
+                "install -Dm644 com.carpeludum.KegaFusion.png /app/share/icons/com.carpeludum.KegaFusion.png",
+                "install -Dm644 -t /app/share/applications/ com.carpeludum.KegaFusion.desktop",
+                "install -Dm644 -t /app/share/metainfo/ com.carpeludum.KegaFusion.metainfo.xml"
+            ],
+            "sources" : [
+                {
+                    "type" : "archive",
+                    "url" : "https://www.carpeludum.com/download/Fusion363x.tar.gz",
```
Comment (gasinvein) [inline com.carpeludum.KegaFusion.json]: Does it's licence allow redistribution? If not (likely so), Fusion should be downloaded at install-time via extra-data.

**Comment by barthalion** [comment]
> Update to 19.08 runtime (there's no i386 support, so how do we compile/run for i386?)

There is i386 compat extension. I think Steam uses it.

**Comment by hadess** [comment]
> There is i386 compat extension. I think Steam uses it.

Took me a while, but I've figured it out :)

Looks like PCSX2 is the only other app in Flathub that uses 32-bit x86 exclusively.

**Comment by hadess** [comment]
Before merging, I need to figure out what to do with the config files and their location, and more testing.

**Comment by barthalion** [comment]
bot, build com.carpeludum.KegaFusion

**Comment by hadess** [comment]
Build fails on screenshots in the appdata. I'll get to those once I've figured out how to handle the configuration file location.

**Comment by hadess** [comment]
> Build fails on screenshots in the appdata. I'll get to those once I've figured out how to handle the configuration file location.

I've now fixed the configuration file location problems. I'll upload some real screenshots once the software is accepted.

**Comment by barthalion** [comment]
bot, build com.carpeludum.KegaFusion

File: com.carpeludum.KegaFusion.desktop (hunk 3 above)
Comment (barthalion) [reply com.carpeludum.KegaFusion.desktop:6]: I think the comment above still applies.

File: com.carpeludum.KegaFusion.yml [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+      - install -Dm644 -t /app/share/applications/ com.carpeludum.KegaFusion.desktop
+      - install -Dm644 -t /app/share/metainfo/ com.carpeludum.KegaFusion.metainfo.xml
+      # Directories for the extensions
+      - mkdir -p /app/lib/i386-linux-gnu•
+      - mkdir -p /app/lib/debug/lib/i386-linux-gnu•
+      - mkdir -p /app/lib/i386-linux-gnu/GL•
+      - install -Dm644 -t /app/etc ld.so.conf
+
+    sources:
+      - type: archive
```
Comment (barthalion) [inline com.carpeludum.KegaFusion.yml:84]: I think the question here remains whether it's redistributable, but if it's not being actively developed, I guess it can stay that way.

Comment (hadess) [reply com.carpeludum.KegaFusion.yml:84]: I don't have an answer. I'm fairly certain it is redistributable, and has been redistributed, and I'm trying to get in touch with the original author.

Comment (hadess) [reply com.carpeludum.KegaFusion.yml:84]: The developer seems to be fine the software being distributed without restrictions. The software hasn't been updated since 2009, so I doubt we'll ever see an update with a license file for example.

**Comment by nedrichards** [comment]
bot, build com.carpeludum.KegaFusion

File: com.carpeludum.KegaFusion.yml [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+      - type: file
+        path: fix-config-location.c
+
+      - type: file
+        path: directories.h
+
+      - type: file
+        url: http://trya.alwaysdata.net/linux/icons/kega-fusion.png
+        sha256: 13ce738c8e10f92735085f624651a7c983093c785b40f4c2e4c7aee6506fe10b
+        dest-filename: com.carpeludum.KegaFusion.png
```
Comment (hfiguiere) [inline com.carpeludum.KegaFusion.yml:107]: shouldn't the file with that name in this repository be removed?

**Comment by barthalion** [comment]
bot, build com.carpeludum.KegaFusion

**Comment by barthalion** [comment]
@hadess No complaints about the manifest from me, but it just hangs for me after I tried to load a SMS rom.

**Comment by hadess** [comment]
> @hadess No complaints about the manifest from me, but it just hangs for me after I tried to load a SMS rom.

I've tried a couple of different SMS ROMs, and they all seem to work fine.

I had trouble getting the app to even run when I realised that `org.freedesktop.Platform.Compat.i386//21.08` wasn't installed. I might have to add a check for that in the startup script in the future if it becomes a problem.

**Comment by barthalion** [comment]
bot, build com.carpeludum.KegaFusion

> I had trouble getting the app to even run when I realised that org.freedesktop.Platform.Compat.i386//21.08 wasn't installed. I might have to add a check for that in the startup script in the future if it becomes a problem.

This shouldn't be a problem once it's available in the repo, as flatpak will install it automatically. It doesn't work now because flatpak won't install extensions automatically across remotes.
